File: neoforge-26.1.2/src/main/java/com/example/horsegenetics/neoforge/client/GeneIcons.java
package com.example.horsegenetics.neoforge.client;

import com.example.horsegenetics.common.genetics.Gene;
import com.example.horsegenetics.neoforge.HorseGenetics;
import net.minecraft.client.Minecraft;
import net.minecraft.resources.Identifier;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

/**
 * <b>The baked picture of a gene</b> - the same one-horse icon its wiki page and
 * the horse designer's hover card show - as a GUI texture.
 *
 * <p>The icons are baked by {@code :common:bakeGeneIcons} into
 * {@code wiki/assets/gene-icons/}, and the NeoForge build copies them into the
 * jar under {@code assets/horsegenetics/textures/gui/gene_icons/} rather than the
 * repository carrying them twice. So a gene that has an icon on the wiki has
 * one in game, and one baked since the last build does not - rebuild.
 *
 * <p>A gene with <b>no</b> icon - a drop-in gene, or one that paints nothing and
 * so was never photographed - answers {@code null}, and the caller draws no
 * picture. Asking the texture manager for a missing texture would draw the
 * purple-and-black checkerboard instead, which reads as a bug rather than as
 * "nothing to show". The answer is cached per gene: the resource pack stack
 * does not change while a screen is open, and a hover card is drawn every frame.
 */
public final class GeneIcons {

    private static final Map<String, Optional<Identifier>> CACHE = new HashMap<>();

    private GeneIcons() {
    }

    /** The icon texture for {@code gene}, or {@code null} if this build has none. */
    public static Identifier iconFor(Gene gene) {
        return CACHE.computeIfAbsent(gene.key(), GeneIcons::find).orElse(null);
    }

    private static Optional<Identifier> find(String key) {
        String slug = key.substring(key.indexOf('.') + 1);
        Identifier id;
        try {
            id = Identifier.fromNamespaceAndPath(HorseGenetics.MOD_ID, "textures/gui/gene_icons/" + slug + ".png");
        } catch (RuntimeException invalid) {
            return Optional.empty();   // a drop-in gene whose key is not a valid path
        }
        return Minecraft.getInstance().getResourceManager().getResource(id).isPresent()
                ? Optional.of(id) : Optional.empty();
    }
}

File: wiki/horse-designer/js/gene-icons.js
// The baked one-horse picture of a gene - wiki/assets/gene-icons/<slug>.png,
// written by :common:bakeGeneIcons - for the gene hover cards.
//
// Shared by the horse designer (drawn onto its canvas, the twin of
// CustomHorseSpawnScreen's blurb) and the breed designer (an <img> in an HTML
// card). The game side is neoforge/client/GeneIcons, which reads the same PNGs
// out of the jar.
//
// A gene with no icon - one that paints nothing and so was never photographed,
// or a drop-in gene - is remembered as missing, and the card draws no picture
// rather than a broken-image box. Nothing here is genetic; it is a cache of
// Image objects keyed by gene key.
window.HG = window.HG || {};
(function (HG) {
  "use strict";

  var cache = {};   // gene key -> { img, ok, done }

  /** The URL of a gene's icon, relative to a page one folder below wiki/. */
  function url(key, base) {
    var slug = key.slice(key.indexOf(".") + 1);
    return (base || "../assets/gene-icons/") + slug + ".png";
  }

  function entry(key, base) {
    var e = cache[key];
    if (e) return e;
    e = cache[key] = { img: new Image(), ok: false, done: false };
    e.img.onload = function () { e.ok = true; e.done = true; };
    e.img.onerror = function () { e.done = true; };
    e.img.src = url(key, base);
    return e;
  }

  HG.geneIcons = {
    /** The loaded Image, or null while it loads and for good if there is none. */
    image: function (key, base) {
      var e = entry(key, base);
      return e.ok ? e.img : null;
    },
    /** The URL, for an <img>. Pair with an onerror that hides it. */
    url: url
  };
})(window.HG);
